File: IsraelHiking.Web/android/app/src/main/java/com/mapeak/car/CarContourTilesProvider.kt
package com.mapeak.car

import com.mapeak.maplibrecontour.DemManager
import com.mapeak.maplibrecontour.DemTileFetcher
import com.mapeak.maplibrecontour.Encoding
import com.mapeak.maplibrecontour.defaultConfig
import com.mapeak.maplibrecontour.parseThresholdSpec
import java.util.concurrent.ConcurrentHashMap
import okhttp3.OkHttpClient
import okhttp3.Request

/**
 * Generates vector contour tiles on the fly from a terrain (DEM) source, mirroring what
 * maplibre-contour does on the web. The heavy lifting is done by the maplibre-contour-rs native
 * library: we hand it a [DemTileFetcher] that knows how to download a DEM tile, and it returns a
 * ready-to-render MVT for any `z/x/y`. [SliceProtocolInterceptor] is what routes contour tile
 * requests here, passing the units carried in the tile URL.
 *
 * Units (metric/imperial) only change the elevation multiplier, so a manager is built and cached per
 * units string; the managers are thread-safe and shared across requests.
 */
class CarContourTilesProvider internal constructor(private val okHttpClient: OkHttpClient) {

    // The layer/attribute names and thresholds must match what the style's contour layers expect
    // (see the web useContourProtocol in default-style.service.ts, which configures the same values).
    private val config =
            defaultConfig()
                    .copy(
                            encoding = Encoding.TERRARIUM,
                            demUrlPattern =
                                    "https://global.israelhikingmap.workers.dev/jaxa_terrarium0-11_v2/{z}/{x}/{y}.webp?use=slice",
                            demMaxZoom = 11.toUByte(),
                            overzoom = 1.toUByte(),
                            thresholds =
                                    parseThresholdSpec(
                                            "11*200*1000~12*10*100~13*10*100~14*10*100~15*10*100"
                                    ),
                            layerName = "contours",
                            elevationKey = "ele",
                            levelKey = "level",
                    )

    // Called by the native library to download the raw DEM tile bytes for a fully-resolved URL.
    // okHttpClient carries SliceProtocolInterceptor (the `use=slice` marker in demUrlPattern), so
    // DEM fetches get the same offline (PMTiles) fallback the rest of the map enjoys. Returning null
    // tells the library there is no data for that tile, which it renders as an empty contour tile.
    private val fetcher =
            object : DemTileFetcher {
                override fun fetch(url: String): ByteArray? {
                    val request = Request.Builder().url(url).build()
                    okHttpClient.newCall(request).execute().use { response ->
                        if (!response.isSuccessful) {
                            return null
                        }
                        return response.body.bytes()
                    }
                }
            }

    // One manager per units string; metric and imperial differ only by the elevation multiplier.
    private val managers = ConcurrentHashMap<String, DemManager>()

    fun getTile(z: Int, x: Int, y: Int, units: String): ByteArray {
        val manager =
                managers.computeIfAbsent(units) {
                    DemManager(fetcher, config.copy(multiplier = multiplierForUnits(it)))
                }
        return manager.tile(z.toUByte(), x.toUInt(), y.toUInt())
    }

    companion object {
        private const val UNIT_IMPERIAL = "imperial"
        private const val METRIC_MULTIPLIER = 1.0f
        private const val IMPERIAL_MULTIPLIER = 3.28084f

        private fun multiplierForUnits(units: String): Float =
                if (units == UNIT_IMPERIAL) IMPERIAL_MULTIPLIER else METRIC_MULTIPLIER
    }
}

File: IsraelHiking.Web/ios/App/App/Car/CarContourTilesProvider.swift
import Foundation
import MaplibreContour

/**
 * Generates vector contour tiles on the fly from a terrain (DEM) source. The native
 * maplibre-contour-rs library does the heavy lifting: we hand it a `DemTileFetcher` that downloads a
 * DEM tile and it returns a ready-to-render MVT for any z/x/y.
 *
 * Units (metric/imperial) only change the elevation multiplier, so a manager is built and cached per
 * units string; the managers are thread-safe and shared across requests.
 */
final class CarContourTilesProvider {

    private let baseConfig: ContourConfig
    private let fetcher: DemTileFetcher
    private let lock = NSLock()
    private var managers: [String: DemManager] = [:]

    init() {
        var config = defaultConfig()
        config.encoding = .terrarium
        config.demUrlPattern =
            "https://global.israelhikingmap.workers.dev/jaxa_terrarium0-11_v2/{z}/{x}/{y}.webp?use=slice"
        config.demMaxZoom = 11
        config.overzoom = 1
        config.thresholds = parseThresholdSpec(spec: "11*200*1000~12*10*100~13*10*100~14*10*100~15*10*100")
        config.layerName = "contours"
        config.elevationKey = "ele"
        config.levelKey = "level"
        baseConfig = config
        fetcher = ContourDemTileFetcher()
    }

    /// Returns the contour MVT for the tile, generated with the multiplier matching `units`.
    func getTile(z: Int, x: Int, y: Int, units: String) throws -> Data {
        try manager(for: units).tile(z: UInt8(z), x: UInt32(x), y: UInt32(y))
    }

    /// Retrieves a DemManager instance for the given units, creating one if necessary.
    private func manager(for units: String) -> DemManager {
        lock.lock(); defer { lock.unlock() }
        if let cached = managers[units] { return cached }
        var config = baseConfig
        config.multiplier = Self.multiplier(for: units)
        let manager = DemManager(fetcher: fetcher, config: config)
        managers[units] = manager
        return manager
    }

    private static func multiplier(for units: String) -> Float {
        units == unitImperial ? imperialMultiplier : metricMultiplier
    }

    private static let unitImperial = "imperial"
    private static let metricMultiplier: Float = 1.0
    private static let imperialMultiplier: Float = 3.28084
}

/**
 * Downloads the raw DEM tile bytes for a fully-resolved URL. The `use=slice` marker in the DEM URL
 * pattern routes the request through `SliceURLProtocol`, giving DEM fetches the same offline
 * (PMTiles) fallback the rest of the map enjoys. Returns nil when there is no data, which the library
 * renders as an empty contour tile.
 */
private final class ContourDemTileFetcher: DemTileFetcher, @unchecked Sendable {

    private let session: URLSession

    init() {
        let config = URLSessionConfiguration.ephemeral
        var protocols: [AnyClass] = [SliceURLProtocol.self]
        protocols.append(contentsOf: config.protocolClasses ?? [])
        config.protocolClasses = protocols
        session = URLSession(configuration: config)
    }

    /// Called synchronously by the native library; block until the tile request completes.
    func fetch(url: String) -> Data? {
        guard let requestUrl = URL(string: url) else { return nil }
        let semaphore = DispatchSemaphore(value: 0)
        var result: Data?
        let task = session.dataTask(with: requestUrl) { data, response, _ in
            if let http = response as? HTTPURLResponse, http.statusCode == 200 {
                result = data
            }
            semaphore.signal()
        }
        task.resume()
        semaphore.wait()
        return result
    }
}

